### PR TITLE
Make hooks funcallable.

### PR DESCRIPTION
**Pull request by aartaka**
This makes hooks `funcall`-able. Not much practical value, except allowing for more relaxed hook invocations, compared to the regular `nhooks:run-hook`.

This adds a `closer-mop` dependency, which might be problematic for other projects (potentially) relying on this library, but it's fine for Nyxt—it already depends on `closer-mop` :)

As an additional mind-boggling thing: SBCL allows [extending `sequence` class](http://www.sbcl.org/manual/index.html#Extensible-Sequences) to allow for any class to behave like sequences. Maybe implement this too, indexing handlers etc...

Does that look sane?

### CODE REVIEW

**Comment by aartaka** [comment]
@Ambrevar, any feedback?

File: nhooks.lisp [hunk 1]
```diff
@@ -251,6 +253,9 @@ handlers-alist."
     (handlers-alist hook)))
 
 (defmethod run-hook ((hook hook) &rest args)
+  "Invoke all the HOOK handlers with the default `combination'.
+
+Alternatively, use `funcall' of the hook for the same effect."
```
Comment (Ambrevar) [inline nhooks.lisp:258]: Typo: "on the hook" or just "`funcall' the hook".

**Review by Ambrevar — COMMENTED** [review]
Looks good to me.

What about making handlers funcallable too?

**Comment by Ambrevar** [comment]
> This adds a `closer-mop` dependency, which might be problematic for other projects (potentially) relying on this library

`closer-mop` could have been named `trivial-mop`: it's super lightweight, no dependencies.  So no hassle, we can include it!

**Comment by aartaka** [comment]
OK, handlers are funcallable now. Which makes me wonder: do we actually need `fn` anymore?—it all can be simulated with `funcall`-able instance setting. Not sure how symbol-based handlers would work, but we can generate proxy lambdas for those in the worst case ( ͡° ͜ʖ ͡°)

File: nhooks.lisp [hunk 2]
```diff
@@ -38,14 +38,18 @@ If the handler is meant to be a setter, VALUE can be used to describe what FN is
...
 
 - They can embed a NAME so that anonymous functions can be conveniently used in hooks.
 - If the handler is meant to be a setter, the PLACE and VALUE slots can be used
   to identify and compare setters.
 
-With this extra information, it's possible to compare handlers and, in particular, avoid duplicates in hooks."))
+With this extra information, it's possible to compare handlers and, in particular, avoid duplicates in hooks.
+
+Handlers are `funcall'able. If you subclass those, don't forget to add
+a `closer-mop:funcallable-standard-class' to retain this property."))
```
Comment (Ambrevar) [inline nhooks.lisp:52]: No need to mention this, subclasses do not have a choice, they _must_ use a compatible metaclass and the default one won't do!  Try it, the Lisp will throw an error.

**Comment by Ambrevar** [comment]
I'm not sure we can get rid of `fn` because how would we check the types otherwise?

File: nhooks.lisp (hunk 2 above)
Comment (aartaka) [reply nhooks.lisp:52]: The errors I've got were quite cryptic, so it wouldn't hurt mentioning the exact thing.

**Comment by aartaka** [comment]
> I'm not sure we can get rid of `fn` because how would we check the types otherwise?

We can do that the same way we do it now: by compiling and `ftype`-declaring it. For example, the following two snippets work as intended (on SBCL, CCL, and ECL):
``` lisp
;; Basic type check. Handlers are functions.
(typep (make-instance 'nhooks:handler :name 'hey :fn (lambda () (print "hey!"))) 'function)
;; A more complicated check from nhooks::probe-ftype. Handlers inherit the type of their funcallable instances.
(compile 
 nil
 `(lambda ()
    (let ((fn ,(make-instance 'nhooks:handler :name 'hey :fn (lambda () (print "hey!")))))
      (declare (type (function ()) fn))
      fn)))
```

**Comment by aartaka** [comment]
CL is so well-planned...

**Comment by Ambrevar** [comment]
I wasn't sure it also works on values... Just checked: it does.

From what I recall it does not work on CCL because the compiler does less function type checking, plus it does not raise a condition.

Did you get it to work on CCL?

Also you are still using `:fn`.  Or do you mean to only use it as an `initialize-instance :after` parameter?

**Comment by aartaka** [comment]
> I wasn't sure it also works on values... Just checked: it does.
> 
> From what I recall it does not work on CCL because the compiler does less function type checking, plus it does not raise a condition.
> 
> Did you get it to work on CCL?

I did, but note that I've checked to arg-less function type, and it may stop working if we add args...

> Also you are still using `:fn`. Or do you mean to only use it as an `initialize-instance :after` parameter?

I am using it, but we can safely turn it into an `initialize-instance :after` parameter and `closer-mop:set-funcallable-instance-function` on it there!

**Comment by Ambrevar** [comment]
Exactly, CCL cannot go as sophisticated as SBCL.  Try the following on various compiler:
```lisp
(compile 
 nil
 `(lambda ()
    (let ((fn (lambda (s) (* 2 s))))
      (declare (type (function (number) string) fn))
      fn)))

(compile 
 nil
 `(lambda ()
    (let ((fn (lambda (s) (declare (type number s)) (* 2 s))))
      (declare (type (function (string) number) fn))
      fn)))
```

**Comment by aartaka** [comment]
But then, does it work right now? What would incorporating functions inside handlers change anyway?

**Comment by Ambrevar** [comment]
Sorry, what works right now?

**Comment by aartaka** [comment]
I mean, handler/hook `ftype` checking on CCL. Your snippet applies to functions, which makes me think that it doesn't work as a super-reliable `ftype` checker, at least on CCL. But then, you've used it as an argument against `:fn` slots being merged inside handlers, somewhy saying that situation with `ftype` checking would change in some way if we merge those.

But it won't -- CCL `ftype` checking would still be too permissive, no matter what we're checking -- functions or some `funcallable-standard-object`. It should be fine merging `fn` inside the `handler`, as it will change nothing except removing an unnecessary slot from the semantically funcallable and `ftype`-check-able object.

**Comment by Ambrevar** [comment]
Misunderstanding: I was just being confused because nhooks used to be implemented differently, by type-checking functions.  But that does not make sense with the current implementation.

All good!

**Comment by aartaka** [comment]
Okay, there's no more `fn`. However, all this funcallability business strips us off the most important feature of `nhooks` -- binding handlers to symbols, with the ability of symbol-bound function redefinition.

Given that, we'd have to revert the handler funcallability so that handlers use the old mechanism. But then, it wouldn't make much sense making hooks solely funcallable, I guess?

I can see the benefits of both approaches (making everything funcallable vs. allowing symbol hooks), but I guess allowing symbols is more pragmatic. @Ambrevar, @BlueFlo0d, @aadcg, any ideas on that? Can we have both?

Unless anyone is strongly against it, or unless we find some way to support symbols, I'll close this PR in a week :)

**Comment by Ambrevar** [comment]
Oh yes, that makes sense.

Hanving funcallable hooks and handlers is still possible even if we keep the `fn` slot though, no?

**Comment by aartaka** [comment]
OK, the old behavior is preseverd (in regards to symbol-bound handlers), while both hooks and handlers are properly funcallable. It's fully backward-compatible. I'll think some more on how to make it more elegant, and will merge today's evening.

**Comment by aadcg** [comment]
Interesting work. Nothing much to add :)

**Comment by Ambrevar** [comment]
Thanks!

**Pull request merged**